xml/map.go

```diff
@@ -0,0 +1,49 @@
+package xml
+
+// mapEntryWrapper is the default member wrapper start element for XML Map entry
+var mapEntryWrapper = StartElement{
+	Name: Name{Local: "entry"},
+}
+
+// Map represents the encoding of a XML map type
+type Map struct {
+	w       writer
+	scratch *[]byte
+
+	// member start element is the map entry wrapper start element
+	memberStartElement StartElement
+
+	// isFlattened returns true if the map is a flattened map
+	isFlattened bool
+}
+
+// newMap returns a map encoder which sets the default map
+// entry wrapper to `entry`.
+//
+// A wrapped map `someMap : {{key:"abc", value:"123"}}` is represented as
+// `<someMap><entry><key>abc<key><value>123</value></entry></someMap>`.
+//
+// A flattened map `someMap : {{key:"abc", value:"123"}}` is represented as
+// `<someMap><key>abc<key><value>123</value></someMap>`.
+func newMap(w writer, scratch *[]byte, memberStartElement StartElement, isFlattened bool) *Map {
+	var memberWrapper = mapEntryWrapper
+	// If flattened map then use member start element as member wrapper
+	if isFlattened {
+		memberWrapper = memberStartElement
+	}
```

Comment:
Instead of having optional parameters, it probably is best to have two different constructor methods.  Optional parameters can easily be confusing.  Since the regular map doesn't use the provided start element its better to have the two constructors.

Regular non-flattened map
```go
func newMap(w writer, scratch *[]byte) *Map
```

Flattened Map
```go
func newMapWithWrapper(w writer, scratch *[]byte, memberWrapper StartElement) *Map {
```